### PR TITLE
Replace `#fileID`, `#filePath`, `#line`, and `#column` with a single macro.

### PR DESCRIPTION
**Pull request by grynspan**
We have a bunch of functions that take separate, defaulted arguments for file ID, column, etc. We'd like to normalize these functions to always take a single `SourceLocation` instance, but until now there hasn't been a way to express such a thing as a default argument without capturing the wrong source location.

This PR takes advantage of [SE-0422](https://github.com/apple/swift-evolution/blob/main/proposals/0422-caller-side-default-argument-macro-expression.md) to create a new `#_sourceLocation` macro (name pending API review) that does what we want here. We can't use the obvious `#sourceLocation` because the language already reserves that macro name for an unrelated use.

Resolves rdar://121883259.

### Checklist:

- [x] Code and documentation should follow the style of the [Style Guide](https://github.com/apple/swift-testing/blob/main/Documentation/StyleGuide.md).
- [x] If public symbols are renamed or modified, DocC references should be updated.

### CODE REVIEW

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

File: Sources/Testing/Testing.docc/ExpectThrows.md
Comment (stmontgomery) [inline Sources/Testing/Testing.docc/ExpectThrows.md:1]: What's the motivation for adding a distinct new article as part of this PR?

Comment (grynspan) [reply Sources/Testing/Testing.docc/ExpectThrows.md:1]: I think that was just a merge failure. Will delete.

File: Sources/Testing/SourceAttribution/SourceLocation+Macro.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+/// Get the current source location as a compile-time constant.
+///
+/// - Returns: The source location at which this macro is applied.
+///
+/// This macro can be used in place of `#fileID`, `#line`, and `#column` as a
+/// default argument to a function. It expands to an instance of
+/// ``SourceLocation`` referring to the location of the macro invocation itself
+/// (similar to how `#fileID` expands to the ID of the file containing the
+/// `#fileID` invocation.)
+@freestanding(expression) public macro currentSourceLocation() -> SourceLocation = #externalMacro(module: "TestingMacros", type: "SourceLocationMacro")
```
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Testing/SourceAttribution/SourceLocation+Macro.swift:20]: Sounds good to me.

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Linux

**Pull request merged**